### PR TITLE
Laboratory work #2, Marina Mokina - 22FPL1

### PR DESCRIPTION
**Pull request by mmarina2004**

### CODE REVIEW

**Comment by artyomtugaryov** [comment]
Nothing to review

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,6 +15,17 @@ def prepare_word(
...
     :return: preprocessed word
     """
+    if not isinstance(raw_word, str) or not (isinstance(
+            start_of_word, str) or start_of_word is None) or not (
+            isinstance(end_of_word, str) or end_of_word is None):
+        return None
+    tokenized_word = []
+    if start_of_word is not None:
+        tokenized_word.append(start_of_word)
+    tokenized_word.extend(element for element in raw_word)
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:25]: strange construction

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -47,6 +80,23 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not (isinstance(word_frequencies, dict)
+            and isinstance(pair, tuple)):
+        return None
+    merged_frequencies = {}
+    for preprocessed_word, count in word_frequencies.items():
+        if ''.join(pair) in ''.join(preprocessed_word):
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:88]: looks like the you can remove `if` and else at all.

File: seminars/practice_3_lists.py [hunk 3]
```diff
@@ -58,11 +58,16 @@ def count_evens(nums: list) -> int:
     """
     Return the number of even ints in the given array.
     """
-    # student realization goes here
+    n = 0
```
Comment (artyomtugaryov) [inline seminars/practice_3_lists.py:61]: revert

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -159,3 +315,24 @@ def calculate_bleu(actual: str | None, reference: str, max_order: int = 3) -> fl
     :param max_order: max length of n-gram to consider for comparison
     :return: value of BLEU metric
     """
+    if not isinstance(actual, str) or not isinstance(reference, str) or not isinstance(max_order, int):
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:318]: I need to see how you use it

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (mmarina2004) [reply lab_2_tokenize_by_bpe/main.py]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (mmarina2004) [reply lab_2_tokenize_by_bpe/main.py:88]: I can only remove else

File: seminars/practice_3_lists.py (hunk 3 above)
Comment (mmarina2004) [reply seminars/practice_3_lists.py]: done

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -47,6 +80,22 @@ def merge_tokens(
...
+    if not (isinstance(word_frequencies, dict)
+            and isinstance(pair, tuple)):
+        return None
+    merged_frequencies = {}
+    for preprocessed_word, count in word_frequencies.items():
+        if ''.join(pair) in ''.join(preprocessed_word):
+            list_word = list(preprocessed_word)
+            for index in range(len(list_word) - 1):
+                if (list_word[index], list_word[index + 1]) == pair:
+                    list_word[index + 1] = pair[0] + pair[1]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:92]: `join(pair)`

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -69,6 +135,20 @@ def get_vocabulary(
...
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(unknown_token, str):
+        return None
+    tokens_list = set()
+    dict_token_identifier = {}
+    for tuples in word_frequencies:
+        for token in tuples:
+            tokens_list.add(token)
+            for element in token:
+                tokens_list.add(element)
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:146]: .update for set

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (mmarina2004) [reply lab_2_tokenize_by_bpe/main.py]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (mmarina2004) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
Comment (mmarina2004) [reply lab_2_tokenize_by_bpe/main.py]: fixed

**Pull request closed without merging**